DESIGN.md

```diff
@@ -0,0 +1,200 @@
+
+We would like a way for element authors to define their own collections controll exactly what elements are included. For this it would be ideal to have a centralized definition of a collection. Even though a collection is not published to npm, a collection definition could be published to a well-known URL, such as a file stored in GitHub. A custom element manifest can then include references to collection definitions so that while indexing element we can discover and index collections.
+
+### Schema
+
+The catalog server has two schemas in play: A GraphQL schema and an implcit Firestore schema (implicit, since Firestore is schemaless). We will try to keep these as close as possible. Some places where schemas differ:
+* IDs: Firestore documents have immutable IDs that must be unique within their subcollection. This makes immutable package data like package name, or the version string for a specific version, well-suited to be document IDs. They will name to GraphQL fields like name or version.
+* Collections: Firestore has both array-valued fields and subcollections. In GraphQL these are both described by a array-valued field, so Firestore subcollections will be mapped to GraphQL array fields.
+* Dates vs Timestamps: Firestore has a Timestamp data type that will be converted to JS dates for GraphQL
+* Maps: Firestore supports map-valued fields. GraphQL only supports list values, so we must convert maps to lists of key/value pairs.
+* Denormalized fields in Firestore documents may or may not be present in the GraphQL schema.
+
+Firestore stores data in documents and collections. Collections and documents have immutable IDs, and can be referenced by a path of alternating collection IDs and document IDs. This forms a hierarchy somewhat like a filesystem.
+
+GraphQL schemas describe a graph of types, though query results are always a tree. GraphQL fields with a list type can represent Firestore collections. Since Firestore is schemaless we'll use the GraphQL schema types to describe Firestore documents.
+
+#### Package, PackageVersion, and CustomElement
+
+One root Firestore collection is `packages`, containing `Package` documents. A `Package` represents a single package name and all published versions of the package, under the subcollection `versions`. `PackageVersion` describes a single publisend version of a package. `PackageVersion` then contains a `customElements` collection of `CustomElement` documents.
+
+We end up with a hierarchy like:
+
+`Package` -> `versions` -> `PackageVersion` -> `customElements` -> `CustomElement`
+
+Only one of the `PackageVersion` documents represents the latest version of a package. Queries will generally be performed against the lastest versions. Other versions will be available for documentation.
+
+#### Collections
+
+Since ollections don't neccesaarily correspond to packages, we will have a separate root Firestore collection to represent element collections.
```

Comment:
```suggestion
Since collections don't neccesaarily correspond to packages, we will have a separate root Firestore collection to represent element collections.
```